summary/task/migrationrunner.go

```diff
@@ -75,27 +74,65 @@ func (r *MigrationRunner) GetRunnerDurationMaximum() time.Duration {
 	return MigrationTaskDurationMaximum
 }
 
-func (r *MigrationRunner) GenerateNextTime(interval MinuteRange) time.Duration {
-	Min := time.Duration(interval.Min) * time.Second
-	Max := time.Duration(interval.Max) * time.Second
+func (r *MigrationRunner) Run(ctx context.Context, tsk *task.Task) {
+	ctx = auth.NewContextWithServerSessionTokenProvider(ctx, r.AuthClient())
+	if taskRunner, err := NewMigrationTaskRunner(r, tsk); err != nil {
+		log.LoggerFromContext(ctx).WithError(err).Warn("Unable to create task runner")
+	} else {
+		taskRunner.Run(ctx)
+	}
+}
+
+type MigrationTaskRunner struct {
+	Provider
+	task     *task.Task
+	context  context.Context
+	logger   log.Logger
+	deadline time.Time
+	config   Configuration
+}
+
+func NewMigrationTaskRunner(provider Provider, tsk *task.Task) (*MigrationTaskRunner, error) {
+	if provider == nil {
+		return nil, errors.New("provider is missing")
+	}
+	if tsk == nil {
+		return nil, errors.New("task is missing")
+	}
+
+	return &MigrationTaskRunner{
+		Provider: provider,
+		task:     tsk,
+	}, nil
+}
+
+func (t *MigrationTaskRunner) Run(ctx context.Context) {
+	t.context = ctx
+	t.logger = log.LoggerFromContext(t.context)
+	t.deadline = time.Now().Add(t.GetRunnerDurationMaximum())
+	t.config = t.GetConfig()
 
-	randTime := time.Duration(rand.Int63n(int64(Max - Min + 1)))
-	return Min + randTime
+	t.task.ClearError()
+	if err := t.run(); err == nil {
+		t.rescheduleTask()
+	} else if !t.task.HasError() {
+		t.rescheduleTaskWithResourceError(err)
+	}
 }
 
-func (r *MigrationRunner) GetConfig(tsk *task.Task) TaskConfiguration {
-	var config TaskConfiguration
+func (t *MigrationTaskRunner) GetConfig() Configuration {
+	var config Configuration
```

Comment:
Let's clean this up and have a single level of config properties under data similar to how dexcom configuration is stored.